### PR TITLE
Added recovery command for categories

### PR DESCRIPTION
**Pull request by turbo-ele**
Nested tree of categories can now be recovered by command line:
```
app/console sulu:categories:recover -d --force
```

where `--force` forces execution of recovery and
`-d`also fixes wrong depth of tree

TODO: 
* [x] find depths that have no parents and beginn at depth < 0

### CODE REVIEW

File: src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php [hunk 1]
```diff
@@ -0,0 +1,200 @@
+<?php
+
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php]: missing version

File: src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+/**
+ * @package Sulu\Bundle\CategoryBundle\Command
+ */
+class RecoverCommand extends ContainerAwareCommand
+{
+    protected function configure()
+    {
+        $this->setName('sulu:categories:recover')
+            ->addOption(
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php]: description for command is missing

File: src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            );
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $em = $this->getEntityManager();
+        $force = $input->getOption('force');
+        $fixDepth = $input->getOption('depth');
+
+//        $import = $this->getContainer()->get('sulu_contact.import');
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php]: remove unused code(=

File: src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+        $repo = $this->getCategoryRepository();
+        $verify = $repo->verify();
+
+        $success = false;
+
+        // fix nested tree
+        if ($verify !== true) {
+            $output->writeln(sprintf('<comment>%s errors were found. Repair with --force</comment>', count($verify)));
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php]: write this force-hint only if force is not set !

File: src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        // get categories that have no parent but depth > 0
+        $qb = $this->getCategoryRepository()->createQueryBuilder('c2')
+            ->select('count(c2.id)')
+            ->leftJoin('c2.parent', 'c1')
+            ->where('c2.depth <> 0 AND c2.parent IS NULL');
+
+        return $qb->getQuery()->getSingleScalarResult();
+    }
+
+
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php]: remove empty line

File: src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+use Doctrine\ORM\Query\ResultSetMapping;
+use Sulu\Bundle\CategoryBundle\Entity\CategoryRepository;
+use Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * @package Sulu\Bundle\CategoryBundle\Command
```
Comment (dantleech) [inline src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php]: `@package` is not required

File: src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+                'force',
+                'f',
+                InputOption::VALUE_NONE,
+                'update tree'
+            )
+            ->addOption(
+                'depth',
+                'd',
+                InputOption::VALUE_NONE,
+                'fix depth'
```
Comment (dantleech) [inline src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php]: First letter should be a capital

Comment (dantleech) [inline src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php]: What does this do? I think the option should be called `fix-depth` and the description should be more descriptive :)

File: src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+        return false;
+        
+//        // correct DQL
+//        $dql = "UPDATE SuluCategoryBundle:Category c2
+//                JOIN c2.parent c1
+//                SET c2.depth = c1.depth + 1
+//                WHERE (c2.depth - 1) <> c1.depth";
+//        $query = $this->getEntityManager()->createQuery($dql);
+//        $result = $query->execute();
+//        return $result;
```
Comment (dantleech) [inline src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php]: Removed commented code

File: src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+        $sql = "UPDATE ca_categories c2
+                JOIN ca_categories c1 ON c2.idCategoriesParent = c1.id
+                SET c2.depth = (c1.depth + 1)
+                WHERE ( c2.depth - 1 ) <> c1.depth";
+        
+        $statement = $this->getEntityManager()->getConnection()->prepare($sql);
+        if ($statement->execute()) {
+            return $statement->rowCount();
+        }
+        return false;
```
Comment (dantleech) [inline src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php]: By convention there is always an empty line before a return statement.

File: src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+            ->where('c2.depth <> 0 AND c2.parent IS NULL');
+
+        return $qb->getQuery()->getSingleScalarResult();
+    }
+
+
+    /**
+     * fix categories where difference to parents depth
+     *
+     * @return object
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Command/RecoverCommand.php]: you does not return an object here

**Comment by wachterjohannes** [comment]
@turbo-ele code is now ok please add a changelog line

**Comment by turbo-ele** [comment]
see first file

**Comment by wachterjohannes** [comment]
@turbo-ele oh overseen (-:

**Pull request merged**